### ISSUE TITLE
difficulty replicating layers with non-visible outputs (e.g. Dropout) with python CustomOp

### ISSUE THREAD

**Issue by erikbeall** [apache/incubator-mxnet#7075]
I'm trying to replicate existing layers in python (the goal is to modify layers, initially in python), and am having particular difficulty with layers that hold onto information from training, such as the mask in Dropout. Is this possible with CustomOp? For example, in mxnet/src/operator/dropout-inl.h, two outputs are created, with the second output being the mask used for dropping out connections. This second output is then used in backprop so the right weights are used. Naively, a python layer for Dropout would assign a mask of indices to the second output, like so:
`self.assign(out_data[1], req[0], mx.nd.array(~indices))`
The trouble with CustomOp is it does not take into account that we only want one output to go on to the next layer, so the bind will fail if I add a second output naively in my CustomOp. 

In C/C++ mxnet, this is handled by the Dropout class implementing a member function for NumVisibleOutputs() to return 1, and NumOutputs() to return 2.  What is the right way to do this in the python CustomOp, or is the only way at present to use an auxiliary output to hold it?

`class Dropout(mx.operator.CustomOp):
    def __init__(self, keep):
        self.keep = float(keep)
        self.drop = 1 - self.keep
        self.scale = 1.0/self.keep
        self.is_train=True
        super(Dropout, self).__init__()
    def forward(self, is_train, req, in_data, out_data, aux):
        x = in_data[0].asnumpy()
        # if we're in training stage, drop out randomly, otherwise, don't drop anything
        if (is_train):
            self.keep_idx = np.random.rand(*x.shape) < self.keep
            self.is_train=True
            x[~keep_idx] = 0
            self.assign(out_data[0], req[0], mx.nd.array(x*self.scale))
            self.assign(out_data[1], req[0], mx.nd.array(keep_idx))
        else:
            self.is_train=False
            self.assign(out_data[0], req[0], mx.nd.array(x))
    def backward(self, req, out_grad, in_data, out_data, in_grad, aux):
        y = out_grad[0].asnumpy()
        keep_idx=out_data[1].asnumpy()
        if (self.is_train==True):
            y[~keep_idx]=0
            self.assign(in_grad[0], req[0], mx.nd.array(y*self.scale))
        else:
            self.assign(in_grad[0], req[0], mx.nd.array(y))`

`@mx.operator.register("replicated_dropout")
class DropoutProp(mx.operator.CustomOpProp):
    def __init__(self, keep):
        self.keep = keep
        super(DropoutProp, self).__init__(need_top_grad=True)

    def list_arguments(self):
        return ['data']

    def list_outputs(self):
        #return ['output']
        return ['data', 'mask']

    def infer_shape(self, in_shape):
        data_shape = in_shape[0]
        output_shape = in_shape[0]
        aux_shape = in_shape[0]
        return [data_shape], [output_shape, output_shape], []

    def infer_type(self, in_type):
        dtype = in_type[0]
        return [dtype], [dtype, dtype], []

    def create_operator(self, ctx, shapes, dtypes):
        print('creating Dropout operator')
        return Dropout(self.keep)

net_stock = mx.sym.Dropout(data=net, name='stock_drop', p=0.5)
net_custom = mx.symbol.Custom(data=net, name='custom_drop', op_type='replicated_dropout', keep=0.5)`

**Comment by erikbeall** [comment]
Examples/documentation on writing python operators that are more complex than softmax would be very helpful!

**Comment by piiswrong** [comment]
you can ignore the second output

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!
Also, do please check out our [forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) for general "how-to" questions.
